File: app/src/main/java/dev/ragnarok/fenrir/adapter/fave/FaveLinksAdapter.java
package dev.ragnarok.fenrir.adapter.fave;

import static dev.ragnarok.fenrir.util.Objects.nonNull;

import android.content.Context;
import android.view.ContextMenu;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import dev.ragnarok.fenrir.Constants;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.model.FaveLink;
import dev.ragnarok.fenrir.model.PhotoSizes;
import dev.ragnarok.fenrir.picasso.PicassoInstance;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.util.Utils;
import dev.ragnarok.fenrir.util.ViewUtils;

public class FaveLinksAdapter extends RecyclerView.Adapter<FaveLinksAdapter.Holder> {

    private final Context context;
    private List<FaveLink> data;
    private RecyclerView recyclerView;
    private ClickListener clickListener;

    public FaveLinksAdapter(List<FaveLink> data, Context context) {
        this.data = data;
        this.context = context;
    }

    public String getImageUrl(FaveLink link) {
        if (nonNull(link.getPhoto()) && nonNull(link.getPhoto().getSizes())) {
            PhotoSizes sizes = link.getPhoto().getSizes();
            return sizes.getUrlForSize(Settings.get().main().getPrefPreviewImageSize(), true);
        }
        return null;
    }

    @NonNull
    @Override
    public Holder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new Holder(LayoutInflater.from(context).inflate(R.layout.item_fave_link, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull Holder holder, int position) {
        FaveLink item = data.get(position);
        if (Utils.isEmpty(item.getTitle()))
            holder.tvTitle.setVisibility(View.GONE);
        else {
            holder.tvTitle.setVisibility(View.VISIBLE);
            holder.tvTitle.setText(item.getTitle());
        }
        if (Utils.isEmpty(item.getDescription()))
            holder.tvDescription.setVisibility(View.GONE);
        else {
            holder.tvDescription.setVisibility(View.VISIBLE);
            holder.tvDescription.setText(item.getDescription());
        }
        if (Utils.isEmpty(item.getUrl()))
            holder.tvURL.setVisibility(View.GONE);
        else {
            holder.tvURL.setVisibility(View.VISIBLE);
            holder.tvURL.setText(item.getUrl());
        }

        String imageUrl = getImageUrl(item);
        if (imageUrl != null) {
            holder.ivEmpty.setVisibility(View.GONE);
            holder.ivImage.setVisibility(View.VISIBLE);
            ViewUtils.displayAvatar(holder.ivImage, null, imageUrl, Constants.PICASSO_TAG);
        } else {
            PicassoInstance.with().cancelRequest(holder.ivImage);
            holder.ivImage.setVisibility(View.GONE);
            holder.ivEmpty.setVisibility(View.VISIBLE);
        }

        holder.itemView.setOnClickListener(v -> {
            if (clickListener != null) {
                clickListener.onLinkClick(holder.getBindingAdapterPosition(), item);
            }
        });
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    public void setData(List<FaveLink> data) {
        this.data = data;
        notifyDataSetChanged();
    }

    @Override
    public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
        this.recyclerView = recyclerView;
    }

    @Override
    public void onDetachedFromRecyclerView(@NonNull RecyclerView recyclerView) {
        super.onDetachedFromRecyclerView(recyclerView);
        this.recyclerView = null;
    }

    public void setClickListener(ClickListener clickListener) {
        this.clickListener = clickListener;
    }

    public interface ClickListener {
        void onLinkClick(int index, FaveLink link);

        void onLinkDelete(int index, FaveLink link);
    }

    public class Holder extends RecyclerView.ViewHolder implements View.OnCreateContextMenuListener {

        final ImageView ivImage;
        final ImageView ivEmpty;
        final TextView tvTitle;
        final TextView tvDescription;
        final TextView tvURL;

        public Holder(View root) {
            super(root);
            itemView.setOnCreateContextMenuListener(this);

            ivImage = root.findViewById(R.id.item_fave_link_image);
            ivEmpty = root.findViewById(R.id.item_fave_link_empty);
            tvTitle = root.findViewById(R.id.item_fave_link_title);
            tvDescription = root.findViewById(R.id.item_fave_link_description);
            tvURL = root.findViewById(R.id.item_fave_link_url);
        }

        @Override
        public void onCreateContextMenu(ContextMenu menu, View v, ContextMenu.ContextMenuInfo menuInfo) {
            int position = recyclerView.getChildAdapterPosition(v);
            FaveLink faveLink = data.get(position);
            menu.setHeaderTitle(faveLink.getTitle());

            menu.add(0, v.getId(), 0, R.string.delete).setOnMenuItemClickListener(item -> {
                if (clickListener != null) {
                    clickListener.onLinkDelete(position, faveLink);
                }
                return true;
            });
        }
    }
}

File: app/src/main/kotlin/dev/ragnarok/fenrir/adapter/fave/FaveLinksAdapter.kt
package dev.ragnarok.fenrir.adapter.fave

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.view.ContextMenu
import android.view.ContextMenu.ContextMenuInfo
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnCreateContextMenuListener
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import dev.ragnarok.fenrir.Constants
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.model.FaveLink
import dev.ragnarok.fenrir.picasso.PicassoInstance.Companion.with
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.util.CustomToast.Companion.CreateCustomToast
import dev.ragnarok.fenrir.util.ViewUtils.displayAvatar

class FaveLinksAdapter(private var data: List<FaveLink>, private val context: Context) :
    RecyclerView.Adapter<FaveLinksAdapter.Holder>() {
    private var recyclerView: RecyclerView? = null
    private var clickListener: ClickListener? = null
    private fun getImageUrl(link: FaveLink): String? {
        return link.photo?.sizes?.getUrlForSize(Settings.get().main().prefPreviewImageSize, true)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): Holder {
        return Holder(LayoutInflater.from(context).inflate(R.layout.item_fave_link, parent, false))
    }

    override fun onBindViewHolder(holder: Holder, position: Int) {
        val item = data[position]
        if (item.title.isNullOrEmpty()) holder.tvTitle.visibility = View.GONE else {
            holder.tvTitle.visibility = View.VISIBLE
            holder.tvTitle.text = item.title
        }
        if (item.description.isNullOrEmpty()) holder.tvDescription.visibility = View.GONE else {
            holder.tvDescription.visibility = View.VISIBLE
            holder.tvDescription.text = item.description
        }
        if (item.url.isNullOrEmpty()) holder.tvURL.visibility = View.GONE else {
            holder.tvURL.visibility = View.VISIBLE
            holder.tvURL.text = item.url
        }
        val imageUrl = getImageUrl(item)
        if (imageUrl != null) {
            holder.ivEmpty.visibility = View.GONE
            holder.ivImage.visibility = View.VISIBLE
            displayAvatar(holder.ivImage, null, imageUrl, Constants.PICASSO_TAG)
        } else {
            with().cancelRequest(holder.ivImage)
            holder.ivImage.visibility = View.GONE
            holder.ivEmpty.visibility = View.VISIBLE
        }
        holder.itemView.setOnClickListener {
            clickListener?.onLinkClick(holder.bindingAdapterPosition, item)
        }
    }

    override fun getItemCount(): Int {
        return data.size
    }

    fun setData(data: List<FaveLink>) {
        this.data = data
        notifyDataSetChanged()
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        this.recyclerView = recyclerView
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        super.onDetachedFromRecyclerView(recyclerView)
        this.recyclerView = null
    }

    fun setClickListener(clickListener: ClickListener?) {
        this.clickListener = clickListener
    }

    interface ClickListener {
        fun onLinkClick(index: Int, link: FaveLink)
        fun onLinkDelete(index: Int, link: FaveLink)
    }

    inner class Holder(root: View) : RecyclerView.ViewHolder(root), OnCreateContextMenuListener {
        val ivImage: ImageView
        val ivEmpty: ImageView
        val tvTitle: TextView
        val tvDescription: TextView
        val tvURL: TextView
        override fun onCreateContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenuInfo?) {
            val position = recyclerView?.getChildAdapterPosition(v) ?: 0
            val faveLink = data[position]
            menu.setHeaderTitle(faveLink.title)
            menu.add(0, v.id, 0, R.string.delete).setOnMenuItemClickListener {
                clickListener?.onLinkDelete(position, faveLink)
                true
            }
            menu.add(0, v.id, 0, R.string.copy_url).setOnMenuItemClickListener {
                val clipboard =
                    context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager?
                val clip = ClipData.newPlainText("response", faveLink.url)
                clipboard?.setPrimaryClip(clip)
                CreateCustomToast(context).showToast(R.string.copied)
                true
            }
        }

        init {
            itemView.setOnCreateContextMenuListener(this)
            ivImage = root.findViewById(R.id.item_fave_link_image)
            ivEmpty = root.findViewById(R.id.item_fave_link_empty)
            tvTitle = root.findViewById(R.id.item_fave_link_title)
            tvDescription = root.findViewById(R.id.item_fave_link_description)
            tvURL = root.findViewById(R.id.item_fave_link_url)
        }
    }
}